### PR TITLE
[fix][txn] Fix getting last message ID when there are ongoing transactions

### PR DESCRIPTION
**Pull request by liangyepianzhou**
### Motivation
1. Consumer/Reader API `getLastMessageIds` should return the max read position instead of `lastConfirmedEntry`.
https://github.com/apache/pulsar/blob/2322004069f23ecca1a12a4ced898fed854275fb/pulsar-broker/src/main/java/org/apache/pulsar/broker/service/ServerCnx.java#L2117-L2122
2. When the transaction buffer is recovering, the `MaxReadPosition` of the topic is unknown. Wait it recovers completely is needed.
3. The `MaxReadPosition` should be the previous position of the first ongoing transaction position instead of `EntryID -1`. Because compacted decide whether the topic's managedLedger is null by `EntryID == -1`
https://github.com/apache/pulsar/blob/2322004069f23ecca1a12a4ced898fed854275fb/pulsar-broker/src/main/java/org/apache/pulsar/broker/service/ServerCnx.java#L2161-L2176

### Modifications
1. Return `max read position` when calling `getLastMessageIds`.
2. Call `checkIfTBRecoverCompletely` before getting the max read position.
3. If transaction buffer recovery fails, return `Unknown` error to the client. 
### Verifying this change

- [ ] Make sure that the change passes the CI checks.

*(Please pick either of the following options)*

This change is a trivial rework / code cleanup without any test coverage.

*(or)*

This change is already covered by existing tests, such as *(please describe tests)*.

*(or)*

This change added tests and can be verified as follows:

*(example:)*
  - *Added integration tests for end-to-end deployment with large payloads (10MB)*
  - *Extended integration test for recovery after broker failure*

### Does this pull request potentially affect one of the following parts:

*If the box was checked, please highlight the changes*

- [ ] Dependencies (add or upgrade a dependency)
- [ ] The public API
- [ ] The schema
- [ ] The default values of configurations
- [ ] The threading model
- [ ] The binary protocol
- [ ] The REST endpoints
- [ ] The admin CLI options
- [ ] The metrics
- [ ] Anything that affects deployment

### Documentation

- [ ] `doc` 
- [ ] `doc-required` 
- [x] `doc-not-needed` 
- [ ] `doc-complete` 

### Matching PR in forked repository

PR in forked repository:

### CODE REVIEW

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/service/ServerCnx.java [hunk 1]
```diff
@@ -2104,23 +2104,28 @@ protected void handleGetLastMessageId(CommandGetLastMessageId getLastMessageId)
...
+                        topic,
+                        (PositionImpl) lastPosition,
+                        (PositionImpl) markDeletePosition,
+                        partitionIndex,
+                        requestId,
+                        consumer.getSubscription().getName());
+            }).exceptionally(e -> {
+                writeAndFlush(Commands.newError(getLastMessageId.getRequestId(),
+                        ServerError.ServiceNotReady, "Transaction Buffer is not ready."));
+                return null;
```
Comment (codelipenghui) [inline pulsar-broker/src/main/java/org/apache/pulsar/broker/service/ServerCnx.java:2127]: Not all exceptions are retryable.

**Review by codelipenghui — CHANGES_REQUESTED** [review]

File: pulsar-broker/src/test/java/org/apache/pulsar/broker/transaction/buffer/TopicTransactionBufferTest.java [hunk 2]
```diff
@@ -179,4 +191,40 @@ public void testCloseTransactionBufferWhenTimeout() throws Exception {
...
+                .when(transactionBuffer).getMaxReadPosition();
+        try {
+            consumer.getLastMessageIds();
+            fail();
+        } catch (Exception ignore) {}
+        List<TopicMessageId> messageIdList = consumer.getLastMessageIds();
+        assertEquals(messageIdList.size(), 1);
+        TopicMessageIdImpl actualMessageID = (TopicMessageIdImpl) messageIdList.get(0);
+        assertEquals(messageId.getLedgerId(), actualMessageID.getLedgerId());
+        assertEquals(messageId.getEntryId(), actualMessageID.getEntryId());
```
Comment (codelipenghui) [inline pulsar-broker/src/test/java/org/apache/pulsar/broker/transaction/buffer/TopicTransactionBufferTest.java:228]: We should also cover the case that the transaction buffer have ongoing transactions?
And the PR title should also change to "Fix getting last message ID when there are ongoing transactions"

**Review by codelipenghui — CHANGES_REQUESTED** [review]

**Comment by codelipenghui** [comment]
@coderzc 

Could you please help review

> The MaxReadPosition should be the previous position of the first ongoing transaction position instead of EntryID -1.
Compacted decide whether the topic's managedLedger is null by EntryID == -1

I'm not sure what does it actually means.

**Comment by coderzc** [comment]
> @coderzc
> 
> Could you please help review
> 
> > The MaxReadPosition should be the previous position of the first ongoing transaction position instead of EntryID -1.
> > Compacted decide whether the topic's managedLedger is null by EntryID == -1
> 
> I'm not sure what does it actually means.



> @coderzc
> 
> Could you please help review
> 
> > The MaxReadPosition should be the previous position of the first ongoing transaction position instead of EntryID -1.
> > Compacted decide whether the topic's managedLedger is null by EntryID == -1
> 
> I'm not sure what does it actually means.

I think this change is correct since the first entryID of ongoing maybe 0, and its previous entry can't be directly by entryID-1.

**Comment by coderzc** [comment]
@liangyepianzhou I think this pr also fixes #21456.

File: pulsar-broker/src/test/java/org/apache/pulsar/broker/transaction/buffer/TopicTransactionBufferTest.java (hunk 2 above)
Comment (coderzc) [reply pulsar-broker/src/test/java/org/apache/pulsar/broker/transaction/buffer/TopicTransactionBufferTest.java:228]: +1

**Review by coderzc — APPROVED** [review]

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/service/ServerCnx.java [hunk 3]
```diff
@@ -2104,23 +2104,28 @@ protected void handleGetLastMessageId(CommandGetLastMessageId getLastMessageId)
...
+                        (PositionImpl) lastPosition,
+                        (PositionImpl) markDeletePosition,
+                        partitionIndex,
+                        requestId,
+                        consumer.getSubscription().getName());
+            }).exceptionally(e -> {
+                writeAndFlush(Commands.newError(getLastMessageId.getRequestId(),
+                        ServerError.TransactionComponentLoadFailed, "Transaction Buffer failed to load."));
+                return null;
+            });
```
Comment (coderzc) [inline pulsar-broker/src/main/java/org/apache/pulsar/broker/service/ServerCnx.java:2128]: Only checkIfTransactionBufferRecoverCompletely fails then return TransactionComponentLoadFailed?

**Review by coderzc — APPROVED** [review]

File: pulsar-common/src/main/proto/PulsarApi.proto [hunk 4]
```diff
@@ -229,6 +229,9 @@ enum ServerError {
                          // use this error to indicate that this producer is now permanently
                          // fenced. Applications are now supposed to close it and create a
                          // new producer
+    TransactionComponentLoadFailed = 26; // When TransactionBuffer, TransactionPendingAck
+                                         // init failed with no-retryable exception.
+                                         // All operations related to this component should be failed with this error.
```
Comment (gaoran10) [inline pulsar-common/src/main/proto/PulsarApi.proto:234]:
```suggestion
                                         // All operations related to these components should fail with this error.
```

File: pulsar-broker/src/test/java/org/apache/pulsar/broker/transaction/buffer/TopicTransactionBufferTest.java [hunk 5]
```diff
@@ -179,4 +193,121 @@ public void testCloseTransactionBufferWhenTimeout() throws Exception {
...
+            expectedLastMessageID = (MessageIdImpl) producer.newMessage().send();
+        }
+        TopicMessageIdImpl actualLastMessageID = (TopicMessageIdImpl) consumer.getLastMessageIds().get(0);
+        assertEquals(expectedLastMessageID.getEntryId(), actualLastMessageID.getEntryId());
+        assertEquals(expectedLastMessageID.getLedgerId(), actualLastMessageID.getLedgerId());
+        // 2.2 Case2: send 2 ongoing transactional messages and 2 original messages.
+        // |1:0|1:1|1:2|txn1->1:3|1:4|txn2->1:5|1:6|.
+        // 2.2.1 Last message ID will not change when txn1 and txn2 do not end.
+        // 2.2.2 Last message ID will update to 1:4 when txn1 committed.
+        // 2.2.3 Last message ID will update to 1:6 when txn2 aborted.
```
Comment (gaoran10) [inline pulsar-broker/src/test/java/org/apache/pulsar/broker/transaction/buffer/TopicTransactionBufferTest.java:287]: Could we split these comments into corresponding code lines?

File: pulsar-broker/src/test/java/org/apache/pulsar/broker/transaction/buffer/TopicTransactionBufferTest.java [hunk 6]
```diff
@@ -179,4 +193,121 @@ public void testCloseTransactionBufferWhenTimeout() throws Exception {
...
+
+        // 2. Test last max read position can be required correctly.
+        // 2.1 Case1: send 3 original messages. |1:0|1:1|1:2|
+        MessageIdImpl expectedLastMessageID = null;
+        for (int i = 0; i < 3; i++) {
+            expectedLastMessageID = (MessageIdImpl) producer.newMessage().send();
+        }
+        TopicMessageIdImpl actualLastMessageID = (TopicMessageIdImpl) consumer.getLastMessageIds().get(0);
+        assertEquals(expectedLastMessageID.getEntryId(), actualLastMessageID.getEntryId());
+        assertEquals(expectedLastMessageID.getLedgerId(), actualLastMessageID.getLedgerId());
```
Comment (gaoran10) [inline pulsar-broker/src/test/java/org/apache/pulsar/broker/transaction/buffer/TopicTransactionBufferTest.java:282]: Could we add a private method for this check? It appears many times.

**Review by gaoran10 — COMMENTED** [review]
Left some trivial comments.

File: pulsar-common/src/main/proto/PulsarApi.proto [hunk 7]
```diff
@@ -229,6 +229,9 @@ enum ServerError {
                          // use this error to indicate that this producer is now permanently
                          // fenced. Applications are now supposed to close it and create a
                          // new producer
+    TransactionComponentLoadFailed = 26; // When TransactionBuffer, TransactionPendingAck
```
Comment (codelipenghui) [inline pulsar-common/src/main/proto/PulsarApi.proto:232]: If you want to add a new error type to the PulsarApi.proto, it should start with a proposal to clarify the motivation and compatibility.

**Review by BewareMyPower — APPROVED** [review]

**Review by codelipenghui — APPROVED** [review]

**Pull request merged**